stan/math/rev/functor/cvodes_integrator.hpp

```diff
@@ -75,116 +183,174 @@ class cvodes_integrator {
+      for (std::size_t i = 0; i < y0_vars_ + args_vars_; i++) {
+        NV_DATA_S(nv_state_sens_[i]) = &coupled_state_[N_] + i * N_;
+      }
     }
-    if (max_num_steps <= 0) {
-      invalid_argument(function_name, "max_num_steps,", max_num_steps, "",
-                       ", must be greater than 0");
+  }
+
+  ~cvodes_integrator() {
+    SUNLinSolFree(LS_);
+    SUNMatDestroy(A_);
+    N_VDestroy_Serial(nv_state_);
+    if (y0_vars_ + args_vars_ > 0) {
+      N_VDestroyVectorArray_Serial(nv_state_sens_, y0_vars_ + args_vars_);
     }
+  }
 
-    const size_t N = y0.size();
-    const size_t M = theta.size();
-    const size_t S = (initial_var::value ? N : 0) + (param_var::value ? M : 0);
+  /**
+   * Solve the ODE initial value problem y' = f(t, y), y(t0) = y0 at a set of
+   * times, { t1, t2, t3, ... } using the stiff backward differentiation formula
+   * (BDF) solver in CVODES.
+   *
+   * @return std::vector of Eigen::Matrix of the states of the ODE, one for each
+   *   solution time (excluding the initial state)
+   */
+  std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> operator()() {
+    return (*this)(0.0, {});
+  }
 
-    using ode_data = cvodes_ode_data<F, T_initial, T_param>;
-    ode_data cvodes_data(f, y0, theta, x, x_int, msgs);
+  std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> operator()(
+      double rtols, std::vector<double> atols) {
+    std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> y;
 
     void* cvodes_mem = CVodeCreate(Lmm);
     if (cvodes_mem == nullptr) {
       throw std::runtime_error("CVodeCreate failed to allocate memory");
     }
 
-    const size_t coupled_size = cvodes_data.coupled_ode_.size();
-
-    std::vector<std::vector<return_type_t<T_initial, T_param, T_t0, T_ts>>> y;
-    coupled_ode_observer<F, T_initial, T_param, T_t0, T_ts> observer(
-        f, y0, theta, t0, ts, x, x_int, msgs, y);
-
     try {
-      check_flag_sundials(CVodeInit(cvodes_mem, &ode_data::cv_rhs, t0_dbl,
-                                    cvodes_data.nv_state_),
+      check_flag_sundials(CVodeInit(cvodes_mem, &cvodes_integrator::cv_rhs,
+                                    value_of(t0_), nv_state_),
                           "CVodeInit");
 
       // Assign pointer to this as user data
       check_flag_sundials(
-          CVodeSetUserData(cvodes_mem, reinterpret_cast<void*>(&cvodes_data)),
+          CVodeSetUserData(cvodes_mem, reinterpret_cast<void*>(this)),
           "CVodeSetUserData");
 
-      cvodes_set_options(cvodes_mem, relative_tolerance, absolute_tolerance,
-                         max_num_steps);
+      cvodes_set_options(cvodes_mem, relative_tolerance_, absolute_tolerance_,
+                         max_num_steps_);
 
       // for the stiff solvers we need to reserve additional memory
       // and provide a Jacobian function call. new API since 3.0.0:
       // create matrix object and linear solver object; resource
       // (de-)allocation is handled in the cvodes_ode_data
```

Comment:
doc out of date? cvodes_ode_data is not used anymore as I see this.

Reply:
I just deleted this doc. I don't think it tells us much.